src/icrar/leap-accelerate/ms/MeasurementSet.cc

```diff
@@ -32,30 +34,102 @@ namespace icrar
     , m_filepath(filepath)
     , m_readAutocorrelations(readAutocorrelations)
     {
-        m_stations = overrideNStations.is_initialized() ? overrideNStations.get() : m_measurementSet->antenna().nrow();
+        // Check and use unique antennas 
+        casacore::Vector<casacore::Int> a1 = m_msmc->antenna1().getColumn();
+        casacore::Vector<casacore::Int> a2 = m_msmc->antenna1().getColumn();
+        auto a1s = std::set<int32_t>(a1.cbegin(), a1.cend());
+        auto a2s = std::set<int32_t>(a2.cbegin(), a2.cend());
+        std::set_union(a1.cbegin(), a1.cend(), a2.cbegin(), a2.cend(), std::inserter(m_antennas, m_antennas.begin()));
+        
+        if(overrideNStations.is_initialized())
+        {
+            m_stations = overrideNStations.get();
+        }
+        else if(m_antennas.size() != m_measurementSet->antenna().nrow())
+        {
+            LOG(warning) << "ms antennas = " << m_measurementSet->antenna().nrow();
+            LOG(warning) << "unique antennas = " << m_antennas.size();
+            LOG(warning) << "using unique antennas";
+            m_stations = m_antennas.size();
+        }
+        else
+        {
+            m_stations = m_measurementSet->antenna().nrow();
+        }
+
+        Validate();
     }
 
-    MeasurementSet::MeasurementSet(const casacore::MeasurementSet& ms, boost::optional<int> overrideNStations, bool readAutocorrelations)
-    : m_measurementSet(std::make_unique<casacore::MeasurementSet>(ms))
-    , m_msmc(std::make_unique<casacore::MSMainColumns>(*m_measurementSet))
-    , m_msc(std::make_unique<casacore::MSColumns>(*m_measurementSet))
-    , m_readAutocorrelations(readAutocorrelations)
+    void MeasurementSet::Validate() const
     {
-        m_stations = overrideNStations.is_initialized() ? overrideNStations.get() : m_measurementSet->antenna().nrow();
-    }
+        //Stations
+        casacore::Vector<casacore::Int> a1 = m_msmc->antenna1().getColumn();
+        casacore::Vector<casacore::Int> a2 = m_msmc->antenna1().getColumn();
+        auto a1s = std::set<int32_t>(a1.cbegin(), a1.cend());
+        auto a2s = std::set<int32_t>(a2.cbegin(), a2.cend());
+        std::set<std::int32_t> antennas;
+        std::set_union(a1.cbegin(), a1.cend(), a2.cbegin(), a2.cend(), std::inserter(antennas, antennas.begin())); 
```

Comment:
This is exactly the same as above, only inserting into a different container, it could be rewritten as a private method to avoid code repetition.

Reply:
moved to a private method for now